src/polyfills/fetch.js

```diff
@@ -0,0 +1,310 @@
+  }
+
+  /**
+   * Drains the response and returns a promise that resolves with the response
+   * text.
+   * @return {!Promise<string>}
+   */
+  text() {
+    return this.drainText_();
+  }
+
+  /**
+   * Drains the response and returns the JSON object.
+   * @return {!Promise<!JsonObject>}
+   */
+  json() {
+    return /** @type {!Promise<!JsonObject>} */ (
+      this.drainText_().then(parseJson));
+  }
+
+  /**
+   * Drains the response and returns a promise that resolves with the response
+   * ArrayBuffer.
+   * @return {!Promise<!ArrayBuffer>}
+   */
+  arrayBuffer() {
+    return /** @type {!Promise<!ArrayBuffer>} */ (
+      this.drainText_().then(utf8Encode));
+  }
+}
+
+/**
+ * Normalized method name by uppercasing.
+ * @param {string|undefined} method
+ * @return {string}
+ * @private
+ */
+function normalizeMethod(method) {
+  if (method === undefined) {
+    return 'GET';
+  }
+  method = method.toUpperCase();
+  dev().assert(
+      allowedMethods.includes(method),
+      'Only one of %s is currently allowed. Got %s',
+      allowedMethods.join(', '),
+      method
+  );
+  return method;
+}
+
+
+/**
+ * Provides access to the response headers as defined in the Fetch API.
+ * @private Visible for testing.
+ */
+class FetchResponseHeaders {
+  /**
+   * @param {!XMLHttpRequest|!XDomainRequest|!XMLHttpRequestDef} xhr
+   */
+  constructor(xhr) {
+    /** @private @const {!XMLHttpRequest|!XDomainRequest|!XMLHttpRequestDef} */
+    this.xhr_ = xhr;
+  }
+
+  /**
+   * @param {string} name
+   * @return {string}
+   */
+  get(name) {
+    return this.xhr_.getResponseHeader(name);
+  }
+
+  /**
+   * @param {string} name
+   * @return {boolean}
+   */
+  has(name) {
+    return this.xhr_.getResponseHeader(name) != null;
+  }
+}
+/**
+ * Returns instance of Response.
+ * @param {?ResponseBodyInit=} body
+ * @param {!ResponseInit=} init
+ */
+export function Response(body, init = {}) {
+  const lowercasedHeaders = map();
+  const data = Object.assign({
+    status: 200,
+    statusText: 'OK',
+    responseText: (body ? String(body) : ''),
+    /**
+     * @param {string} name
+     * @return {string}
+     */
+    getResponseHeader(name) {
+      return lowercasedHeaders[String(name).toLowerCase()] || null;
+    },
+  }, init);
+
+  init.status = init.status || 200;
+
+  if (isArray(init.headers)) {
+    init.headers.forEach(entry => {
+      const headerName = entry[0];
+      const headerValue = entry[1];
+      lowercasedHeaders[String(headerName).toLowerCase()] =
+          String(headerValue);
+    });
+  } else if (isObject(init.headers)) {
+    for (const key in init.headers) {
+      lowercasedHeaders[String(key).toLowerCase()] =
+          String(init.headers[key]);
+    }
+  }
+  if (init.status) {
+    data.status = parseInt(init.status, 10);
+  }
+  if (init.statusText) {
+    data.statusText = String(init.statusText);
+  }
+
+  return new FetchResponse(/** @type {XMLHttpRequestDef} */(data));
```

Comment:
It's kinda weird that `new Response() instanceof Response` is false. I think the only difficulty to making `Response` be the main class is extracting out the headers from XML?

Reply:
not really, i can make this a sub class of FetchResponse and just do `super(blah)`

Reply:
> Not really

Can you explain?

Reply:
``` 
class Response extends FetchResponse{
  constructor(body, init={}) {
    // all the current logic of functionResponse
   // instead of return new FetchResponse(/** @type {XMLHttpRequestDef} */(data));
   super(/** @type {XMLHttpRequestDef} */(data));
  }
}
```

Reply:
I understand the subclassing. I meant what else is the blocker for just using `Response` besides header parsing?

Reply:
Ah ok, the `fetchPolyfill` is  using FetchReponse as passing xhr in the constructor.
Just using `Response` class would just mean a re-write of that part and some additional code there